### PR TITLE
Foil-1 Geometry Adapter: DSDF Stats → Slice Logit Bias

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Foil-1 DSDF channels carry critical upstream geometry signal — confirmed by the catastrophic failure of DSDF-1 channel dropout (#2156, p_tan +5.6%). The model needs this information, but currently has **no dedicated pathway** to route foil-1 geometry into the slice attention mechanism. The GSB routes using gap/stagger scalars (global geometry), but nothing encodes the foil-1 SHAPE distribution.

**Proposed fix:** Compute per-sample summary statistics (mean, std, skewness, kurtosis) of the 4 foil-1 DSDF channels, pass through a tiny MLP (16→32→slice_num), and inject as an additive bias on slice logits in the first TransolverBlock. This gives the attention routing a "geometry fingerprint" of the upstream foil shape.

Key design: **zero-initialized final layer** so the model starts identical to baseline and gradually learns to use the geometry signal. Inspired by GEPS (Takamoto et al., 2024), AdaIN (Huang & Belongie, 2017), and LoRA's zero-init paradigm.

## Instructions

**Step 1 — Add FoilGeomAdapter module** (near other model components):
```python
class FoilGeomAdapter(nn.Module):
    """Computes foil-1 DSDF distribution statistics and maps to slice logit bias."""
    def __init__(self, dsdf_channels=4, stats_per_channel=4, slice_num=96):
        super().__init__()
        in_dim = dsdf_channels * stats_per_channel  # 16
        self.proj = nn.Sequential(
            nn.Linear(in_dim, 32),
            nn.GELU(),
            nn.Linear(32, slice_num),
        )
        # Zero-init final layer: starts as identity (no bias on slices)
        nn.init.zeros_(self.proj[-1].weight)
        nn.init.zeros_(self.proj[-1].bias)

    def forward(self, x_dsdf_foil1):
        """x_dsdf_foil1: [B, N, 4] — foil-1 DSDF channels across all nodes."""
        eps = 1e-8
        mean = x_dsdf_foil1.mean(dim=1)       # [B, 4]
        std = x_dsdf_foil1.std(dim=1) + eps    # [B, 4]
        centered = (x_dsdf_foil1 - mean.unsqueeze(1)) / std.unsqueeze(1)
        skew = (centered ** 3).mean(dim=1)     # [B, 4]
        kurt = (centered ** 4).mean(dim=1)     # [B, 4]
        stats = torch.cat([mean, std, skew, kurt], dim=-1)  # [B, 16]
        return self.proj(stats)  # [B, slice_num]
```

**Step 2 — Add CLI flags:**
```python
parser.add_argument('--foil1_geom_adapter', action='store_true',
                    help='Enable foil-1 geometry adapter on slice logits')
parser.add_argument('--geom_adapter_scale', type=float, default=0.1,
                    help='Scale factor for geometry adapter bias')
```

**Step 3 — Instantiate in model init** (if `args.foil1_geom_adapter`):
```python
if args.foil1_geom_adapter:
    self.foil_geom_adapter = FoilGeomAdapter(
        dsdf_channels=4, stats_per_channel=4, slice_num=args.slice_num
    )
```

**Step 4 — Inject bias in the FIRST TransolverBlock only.** In `Physics_Attention_Irregular_Mesh.forward()` or `TransolverBlock.forward()`, after computing `slice_logits` but before softmax:
```python
if hasattr(self, 'foil_geom_adapter') or (hasattr(model, 'foil_geom_adapter') ...):
    # x_dsdf_foil1 = x[:, :, 2:6]  — foil-1 DSDF channels (check exact indices)
    geom_bias = model.foil_geom_adapter(x[:, :, 2:6])  # [B, slice_num]
    slice_logits = slice_logits + args.geom_adapter_scale * geom_bias.unsqueeze(1)
```

**Important:** Foil-1 DSDF channels are at `x[:, :, 2:6]` (indices 2,3,4,5). Verify this in the data loading — search for `dsdf` or `DSDF` in the code. The exact channel positions may differ.

**Step 5 — Run 2 configs, 2 seeds each** using `--wandb_group edward-foil1-adapter`:

**Config A — scale=0.1:**
```bash
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/foil1-adapter-s0.1" --wandb_group edward-foil1-adapter \
  --foil1_geom_adapter --geom_adapter_scale 0.1 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(repeat with `--seed 73`)

**Config B — scale=0.3 (stronger geometry signal):**
Same command with `--geom_adapter_scale 0.3` and `--wandb_name "edward/foil1-adapter-s0.3"`. (Repeat with `--seed 73`.)

Run all 4 in parallel. Report p_in, p_oodc, p_tan, p_re per config (2-seed avg).

## Baseline

| Metric | Baseline (2-seed avg) |
|--------|----------------------|
| p_in | 13.05 |
| p_oodc | 7.70 |
| **p_tan** | **28.60** |
| p_re | 6.55 |

Baseline W&B runs: d7l91p0x (seed 42, p_tan=28.9), j9btfx09 (seed 73, p_tan=28.3)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Implementation notes

The PR suggested foil-1 DSDF at `x[:,2:6]`, but after verifying the data layout against `prepare_multi.py`, the actual indices are **`x[:,4:8]`** (after pos(2) and saf(2)). This was corrected in the implementation. The adapter uses the correct 4 foil-1 DSDF channels.

Zero-init ordering was carefully handled: `FoilGeomAdapter` is instantiated after `initialize_weights()` so the zero-init on the final MLP layer survives.

### Metrics

**Config A — scale=0.1 (seeds 42, 73):**
| Split | seed=42 (8ryjbj4w) | seed=73 (1n7jwpzy) | Avg |
|-------|-------------------|-------------------|-----|
| p_in  | 13.4 | 14.4 | **13.9** |
| p_oodc | 7.7 | 8.0 | **7.85** |
| p_tan  | 29.2 | 29.4 | **29.3** |
| p_re   | 6.6  | 6.8  | **6.7**  |

**Config B — scale=0.3 (seeds 42, 73):**
| Split | seed=42 (ex4uj2k8) | seed=73 (lytgamfh) | Avg |
|-------|-------------------|-------------------|-----|
| p_in  | 13.0 | 13.3 | **13.15** |
| p_oodc | 7.8 | 7.7 | **7.75** |
| p_tan  | 28.9 | 29.5 | **29.2**  |
| p_re   | 6.3  | 6.5  | **6.4**   |

**Baseline (2-seed avg):**
| Metric | Baseline |
|--------|---------|
| p_in   | 13.05   |
| p_oodc | 7.70    |
| **p_tan** | **28.60** |
| p_re   | 6.55    |

**Full surface MAE for best run (Config B, seed=42, ex4uj2k8):**
- val_in_dist: Ux=3.13, Uy=0.83, p=13.0
- val_ood_cond: p=7.8
- val_tandem_transfer: p=28.9
- val_ood_re: p=6.3
- Volume (val_in_dist): Ux=0.59, Uy=0.19, p=12.09

Peak memory: 46.0–46.2 GB. Runs completed at ~150 epochs (180 min timeout).

W&B runs: `8ryjbj4w` (A/s42), `1n7jwpzy` (A/s73), `ex4uj2k8` (B/s42), `lytgamfh` (B/s73) — group: `edward-foil1-adapter`

### Commands

```bash
# Config A, seed=42
CUDA_VISIBLE_DEVICES=0 python train.py --agent edward \
  --wandb_name "edward/foil1-adapter-s0.1-seed42" --wandb_group edward-foil1-adapter \
  --foil1_geom_adapter --geom_adapter_scale 0.1 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Config B: add --geom_adapter_scale 0.3; seed 73: change --seed 73
```

### What happened

The foil-1 geometry adapter **did not improve tandem transfer accuracy**. The best result (Config B, seed=42, p_tan=28.9) is still 0.3 worse than baseline (28.60). Neither scale achieved improvement on the primary metric, though Config B shows marginal OOD-Re improvement (avg 6.4 vs 6.55).

The hypothesis was that per-sample DSDF distribution statistics (mean/std/skewness/kurtosis) would give the slice routing mechanism a "geometry fingerprint" for the upstream foil, helping the model learn geometry-conditioned attention patterns. However, the signal appears too coarse — 4 moments per channel captures the global distribution but discards the spatial structure of the geometry. The adapter likely provided only noise relative to the already-rich geometric signal already in x.

Scale=0.3 is marginally better than 0.1, suggesting the bias was too weak at scale=0.1 to have any effect. But 0.3 still wasn't sufficient to overcome the noise from distribution-only statistics.

### Bug report

**Pre-existing: Verification section shows `nan` for Pipeline and Manual denorm MAE surf_p comparisons.** The verification code at the end of training logs "POTENTIAL BUG — Match: NO" but with nan/nan across all my runs (and presumably all runs with the verification feature enabled). This appears unrelated to my changes — I reproduced it identically across all 4 runs. The refinement stats (mean/std/max/min) are also all nan. Worth investigating whether the EMA checkpoint loading or the verification dataloader is producing empty tensors.

Additionally, the "Re values in OOD-Re split (should all be same Re=4.445M)" prints actual log-Re values (11.6–15.4), which are plausible log-scale values for the OOD-Re samples (log(4.4M) ≈ 15.3), not 4.445M. The comment in the verification code may be outdated or the val_ood_re split contains multiple Re values.

### Suggested follow-ups

1. **All-blocks injection instead of first-block only**: The PR used first-block-only injection to avoid disrupting higher-level learned patterns, but it may be too weak to influence routing meaningfully. Injecting into all blocks with learned per-block scales could give more impact.
2. **Per-node spatial features**: The global distribution stats lose spatial geometry. A per-node DSDF embedding (e.g., MLP on the 4 raw DSDF values → a learned channel added to x before the transformer) might capture more relevant geometric information.
3. **Mean + max pooling**: Simpler than 4-moment statistics — try replacing mean/std/skew/kurt with just mean + max per channel (8-dim input instead of 16-dim). Might be less noisy.
4. **Learn the scale**: Replace fixed `geom_adapter_scale` with a learned scalar parameter (initialized to 0.1) so the model can decide how much geometry bias to apply.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2173 because ADVISOR: Closing — both scales regress p_tan. scale=0.1: +2.4%, scale=0.3: +2.1%.

Your analysis is correct: the 4-moment statistics (mean/std/skew/kurt) are too coarse a geometry fingerprint. They capture global DSDF distribution but discard the spatial structure — which IS the geometry. The adapter provided noise relative to the already-rich geometric signal in the raw DSDF channels.

Good catch on the DSDF channel indices (x[:,4:8] not x[:,2:6]) — I'll note this for future experiments.

**Bug report noted:** The nan values in the verification section and the Re comment inconsistency deserve investigation. I'll flag these for the team.

Dead end: per-sample DSDF distribution statistics as geometry conditioning. The moments lose the critical spatial information. Raw per-node DSDF (which the model already sees) is more informative.

Moving you to a genuinely radical direction..
